File: Blind 75/10 - Backtracking/JavaScript/39-combination-sum.js
const { performance } = require('perf_hooks');

class Solution {
  method(candidates, target) {
    let res = [];
    candidates.sort();

    const dfs = (i, numbers, total) => {
      if (i >= candidates.length || total > target) return;
      if (total === target) {
        res.push([...numbers]);
        return;
      }

      numbers.push(candidates[i]);
      dfs(i, numbers, total + candidates[i]);
      numbers.pop();
      dfs(i + 1, numbers, total);
    };

    dfs(0, [], 0);
    return res;
  }

  reference(
    candidates,
    target,
    index = 0,
    combination = [],
    combinations = []
  ) {
    const isBaseCase = target < 0;
    if (isBaseCase) return combinations;

    const isTarget = target === 0;
    if (isTarget) return combinations.push(combination.slice());

    for (let i = index; i < candidates.length; i++) {
      this.backTrack(candidates, target, i, combination, combinations);
    }

    return combinations;
  }

  backTrack = (candidates, target, i, combination, combinations) => {
    combination.push(candidates[i]);
    this.reference(
      candidates,
      target - candidates[i],
      i,
      combination,
      combinations
    );
    combination.pop();
  };

  quantify(testCases, runs = 1e6) {
    const runsArr = Array.from({ length: runs });
    const solStart = performance.now();
    runsArr.map((run, i) => {
      testCases.map((input) => {
        if (i === 0) console.log(this.method(...input));
        else this.method(...input);
      });
    });
    console.log(
      `Runtime for solution: ${(performance.now() - solStart) / 1000}\n`
    );

    const refStart = performance.now();
    runsArr.map((run, i) => {
      testCases.map((input) => {
        if (i === 0) console.log(this.reference(...input));
        else this.reference(...input);
      });
    });
    console.log(
      `Runtime for reference: ${(performance.now() - refStart) / 1000}`
    );
  }
}

const test = new Solution();
const testCases = [
  [[2, 3, 6, 7], 7],
  [[2, 3, 5], 8],
  [[2], 1],
];
test.quantify(testCases);

File: Blind 75/10 - Backtracking/39-combination-sum.py
'''
Given an array of distinct integers candidates and a target integer target, return a list of all unique combinations of candidates where the chosen numbers sum to target. You may return the combinations in any order.

The same number may be chosen from candidates an unlimited number of times. Two combinations are unique if the frequency of at least one of the chosen numbers is different.

It is guaranteed that the number of unique combinations that sum up to target is less than 150 combinations for the given input.
'''
from time import time
from typing import List


class Solution:
    def combinationSum(
        self, candidates: List[int], target: int
    ) -> List[List[int]]:
        candidates.sort()
        res = []

        def dfs(i, subset, total):
            if total == target:
                res.append(subset[:])
                return
            if total > target:
                # Backtracking
                return
            for j in range(i, len(candidates)):
                dfs(j, subset + [candidates[j]], total + candidates[j])

        dfs(0, [], 0)
        return res

    def reference(self, candidates: List[int], target: int) -> List[List[int]]:
        res = []

        def dfs(i, curr, total):
            if total == target:
                res.append(curr.copy())
                return
            if i >= len(candidates) or total > target:
                return

            curr.append(candidates[i])
            dfs(i, curr, total + candidates[i])
            curr.pop()
            dfs(i + 1, curr, total)

        dfs(0, [], 0)
        return res

    def quantify(self, test_cases, runs=50000):
        sol_start = time()
        for i in range(runs):
            for case in test_cases:
                if i == 0:
                    print(self.combinationSum(*case))
                else:
                    self.combinationSum(*case)
        print(f'Runtime for our solution: {time() - sol_start}\n')

        ref_start = time()
        for i in range(0, runs):
            for case in test_cases:
                if i == 0:
                    print(self.reference(*case))
                else:
                    self.reference(*case)
        print(f'Runtime for reference: {time() - ref_start}')


if __name__ == '__main__':
    test = Solution()
    test_cases = [([2, 3, 6, 7], 7), ([2, 3, 5], 8), ([2], 1)]
    test.quantify(test_cases)
